compiler/qsc/src/interpret.rs

```diff
@@ -454,13 +454,57 @@ impl Interpreter {
     fn lower(
         &mut self,
         unit_addition: &qsc_frontend::incremental::Increment,
-    ) -> (Vec<StmtId>, Vec<ExecGraphNode>) {
+    ) -> core::result::Result<(Vec<StmtId>, Vec<ExecGraphNode>), Vec<Error>> {
+        if self.capabilities != RuntimeCapabilityFlags::all() {
+            return self.run_fir_passes(unit_addition);
+        }
         let fir_package = self.fir_store.get_mut(self.package);
-        (
+        Ok((
             self.lowerer
                 .lower_and_update_package(fir_package, &unit_addition.hir),
             self.lowerer.take_exec_graph(),
-        )
+        ))
+    }
+
+    fn run_fir_passes(
+        &mut self,
+        unit: &qsc_frontend::incremental::Increment,
+    ) -> core::result::Result<(Vec<StmtId>, Vec<ExecGraphNode>), Vec<Error>> {
+        let mut fir_store = self.fir_store.clone();
+        let mut lowerer = self.lowerer.clone();
```

Comment:
So we skip updating `self.fir_store` if there are pass errors? I have a feeling this will get us into an inconsistent state but I need to think about it to come up with a repro.

Reply:
This may be related:

```python
import qsharp
qsharp.init(target_profile=qsharp.TargetProfile.Adaptive)
qsharp.eval("operation Foo() : Int { use q = Qubit(); mutable x = 1; if MResetZ(q) == One { set x = 2; } x }")
qsharp.eval("Foo()")
```

Reply:
The incremental compiler `qsc/src/incremental.rs` has this "transactional" nature to it, we check the AST and HIR and only update the store if there are no errors. This prevents us from getting into these half-updated states.

Now that lowering is fallible, I have a feeling that these FIR checks need to occur _before_ the AST and HIR is updated, to avoid inconsistencies between the HIR and FIR. But lowering _depends_ on the HIR being updated. Which creates a bit of a chicken-and-egg problem. I'm not sure how we'll address it, but pretty sure it's important.